macaroons/constraints.go

```diff
@@ -0,0 +1,86 @@
+package macaroons
+
+import (
+	"fmt"
+	"net"
+	"time"
+
+	"gopkg.in/macaroon-bakery.v1/bakery/checkers"
+	macaroon "gopkg.in/macaroon.v1"
+)
+
+// Constraint type adds a layer of indirection over macaroon caveats and
+// checkers.
+type Constraint func(*macaroon.Macaroon)
+
+// AddConstraints returns new derived macaroon by applying every passed
+// constraint and tightening its restrictions.
+func AddConstraints(mac *macaroon.Macaroon, cs ...Constraint) *macaroon.Macaroon {
+	newMac := mac.Clone()
+	for _, constraint := range cs {
+		constraint(newMac)
+	}
+	return newMac
+}
+
+// Each *Constraint function is a functional option, which takes a pointer
+// to the macaroon and adds another restriction to it. For each *Constraint,
+// the corresponding *Checker is provided.
+
+// PermissionsConstraint restricts allowed operations set to the ones
+// passed to it.
+func PermissionsConstraint(ops ...string) func(*macaroon.Macaroon) {
```

Comment:
For completeness, this might be nicer named as `AllowConstraint` and have another similar function that's called `DenyConstraint`.

Reply:
Done in ec250b4. Should it be squashed, or complete commit history is better?